File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/SykefraværshistorikkController.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk;

import java.util.Arrays;
import java.util.List;
import javax.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.OverordnetEnhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Underenhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.EnhetsregisteretClient;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.IngenNæringException;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.publiseringsdatoer.api.PubliseringsdatoerService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.aggregert.AggregertStatistikkDto;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.aggregert.AggregertStatistikkService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.kvartalsvis.KvartalsvisSykefraværshistorikk;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.kvartalsvis.KvartalsvisSykefraværshistorikkService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertLegemeldtSykefraværService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertSykefraværService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertSykefraværshistorikk;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.tilgangskontroll.InnloggetBruker;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.tilgangskontroll.TilgangskontrollService;
import no.nav.security.token.support.core.api.Protected;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@Slf4j
@Protected
@RestController
public class SykefraværshistorikkController {

  private final KvartalsvisSykefraværshistorikkService kvartalsvisSykefraværshistorikkService;
  private final TilgangskontrollService tilgangskontrollService;
  private final EnhetsregisteretClient enhetsregisteretClient;
  private final SummertSykefraværService summertSykefraværService;
  private final AggregertStatistikkService aggregertHistorikkService;
  private final PubliseringsdatoerService publiseringsdatoerService;

  // TODO: Fjern når "aggregert"-endepunktet har blitt tatt i bruk
  private final SummertLegemeldtSykefraværService summertLegemeldtSykefraværService;

  public SykefraværshistorikkController(
      KvartalsvisSykefraværshistorikkService kvartalsvisSykefraværshistorikkService,
      TilgangskontrollService tilgangskontrollService,
      EnhetsregisteretClient enhetsregisteretClient,
      SummertSykefraværService summertSykefraværService,
      AggregertStatistikkService aggregertHistorikkService,
      PubliseringsdatoerService publiseringsdatoerService,
      SummertLegemeldtSykefraværService summertLegemeldtSykefraværService) {
    this.kvartalsvisSykefraværshistorikkService = kvartalsvisSykefraværshistorikkService;
    this.tilgangskontrollService = tilgangskontrollService;
    this.enhetsregisteretClient = enhetsregisteretClient;
    this.summertSykefraværService = summertSykefraværService;
    this.aggregertHistorikkService = aggregertHistorikkService;
    this.publiseringsdatoerService = publiseringsdatoerService;
    this.summertLegemeldtSykefraværService = summertLegemeldtSykefraværService;
  }

  @GetMapping(value = "/{orgnr}/sykefravarshistorikk/kvartalsvis")
  public List<KvartalsvisSykefraværshistorikk> hentSykefraværshistorikk(
      @PathVariable("orgnr") String orgnrStr, HttpServletRequest request) {

    Orgnr orgnr = new Orgnr(orgnrStr);

    InnloggetBruker bruker = tilgangskontrollService.hentBrukerKunIaRettigheter();

    tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
        orgnr, bruker, request.getMethod(), "" + request.getRequestURL());

    Underenhet underenhet = enhetsregisteretClient.hentInformasjonOmUnderenhet(orgnr);
    OverordnetEnhet overordnetEnhet =
        enhetsregisteretClient.hentInformasjonOmEnhet(underenhet.getOverordnetEnhetOrgnr());

    boolean harTilgangTilOverordnetEnhet =
        tilgangskontrollService.hentTilgangTilOverordnetEnhetOgLoggSikkerhetshendelse(
            bruker, overordnetEnhet, underenhet, request.getMethod(), "" + request.getRequestURL());

    if (harTilgangTilOverordnetEnhet) {
      return kvartalsvisSykefraværshistorikkService.hentSykefraværshistorikk(
          underenhet, overordnetEnhet);
    } else {
      return kvartalsvisSykefraværshistorikkService.hentSykefraværshistorikk(
          underenhet, overordnetEnhet.getInstitusjonellSektorkode());
    }
  }

  // TODO: Fjern har vi har gått over til "aggregert"-endepunktet
  @GetMapping(value = "/{orgnr}/sykefravarshistorikk/summert")
  public List<SummertSykefraværshistorikk> hentSummertKorttidsOgLangtidsfraværV2(
      @PathVariable("orgnr") String orgnrStr,
      @RequestParam("antallKvartaler") int antallKvartaler,
      HttpServletRequest request) {

    InnloggetBruker bruker = tilgangskontrollService.hentBrukerKunIaRettigheter();
    tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
        new Orgnr(orgnrStr), bruker, request.getMethod(), "" + request.getRequestURL());
    Underenhet underenhet = enhetsregisteretClient.hentInformasjonOmUnderenhet(new Orgnr(orgnrStr));

    if (antallKvartaler != 4) {
      throw new IllegalArgumentException("For øyeblikket støtter vi kun summering av 4 kvartaler.");
    }

    SummertSykefraværshistorikk summertSykefraværshistorikkVirksomhet =
        summertSykefraværService.hentSummertSykefraværshistorikk(
            underenhet, publiseringsdatoerService.hentSistePubliserteKvartal(), antallKvartaler);

    SummertSykefraværshistorikk summertSykefraværshistorikkBransjeEllerNæring =
        summertSykefraværService.hentSummertSykefraværshistorikkForBransjeEllerNæring(
            underenhet, antallKvartaler);

    return Arrays.asList(
        summertSykefraværshistorikkVirksomhet, summertSykefraværshistorikkBransjeEllerNæring);
  }

  @GetMapping(value = "/{orgnr}/sykefravarshistorikk/legemeldtsykefravarsprosent")
  public ResponseEntity<LegemeldtSykefraværsprosent> hentLegemeldtSykefraværsprosent(
      @PathVariable("orgnr") String orgnrStr, HttpServletRequest request) {
    Orgnr orgnr = new Orgnr(orgnrStr);
    InnloggetBruker brukerMedIaRettigheter = tilgangskontrollService.hentBrukerKunIaRettigheter();
    InnloggetBruker brukerMedMinstEnRettighet = tilgangskontrollService.hentInnloggetBrukerForAlleRettigheter();
    boolean brukerHarSykefraværRettighetTilVirksomhet = brukerMedIaRettigheter.harTilgang(orgnr);
    boolean brukerRepresentererVirksomheten = tilgangskontrollService.brukerRepresentererVirksomheten(
        orgnr);

    tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
        new Orgnr(orgnrStr), brukerMedMinstEnRettighet, request.getMethod(),
        "" + request.getRequestURL());

    Underenhet underenhet;
    try {
      underenhet = enhetsregisteretClient.hentInformasjonOmUnderenhet(orgnr);
    } catch (IngenNæringException e) {
      return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
    }

    if (brukerHarSykefraværRettighetTilVirksomhet) {
      LegemeldtSykefraværsprosent legemeldtSykefraværsprosent =
          summertLegemeldtSykefraværService.hentLegemeldtSykefraværsprosent(
              underenhet, publiseringsdatoerService.hentSistePubliserteKvartal());

      if (legemeldtSykefraværsprosent.getProsent() == null) {
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
      }

      return ResponseEntity.status(HttpStatus.OK).body(legemeldtSykefraværsprosent);
    } else if (brukerRepresentererVirksomheten) {
      LegemeldtSykefraværsprosent legemeldtSykefraværsprosentBransjeEllerNæring =
          summertLegemeldtSykefraværService.hentLegemeldtSykefraværsprosentUtenStatistikkForVirksomhet(
              underenhet, publiseringsdatoerService.hentSistePubliserteKvartal());

      if (legemeldtSykefraværsprosentBransjeEllerNæring.getProsent() == null) {
        return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
      }

      return ResponseEntity.status(HttpStatus.OK)
          .body(legemeldtSykefraværsprosentBransjeEllerNæring);
    }
    log.error("Brukeren har ikke tilgang til virksomhet, men ble ikke stoppet av tilgangssjekk.");
    return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(null);
  }

  @GetMapping("/{orgnr}/v1/sykefravarshistorikk/aggregert")
  public ResponseEntity<AggregertStatistikkDto> hentAggregertStatistikk(
      @PathVariable("orgnr") String orgnr) {

    AggregertStatistikkDto statistikk =
        aggregertHistorikkService.hentAggregertStatistikk(new Orgnr(orgnr)).getOrElseThrow(e -> e);

    return ResponseEntity.status(HttpStatus.OK).body(statistikk);
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/SykefraværshistorikkController.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk

import arrow.core.getOrElse
import jakarta.servlet.http.HttpServletRequest
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Underenhet
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.EnhetsregisteretClient
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.EnhetsregisteretClient.HentEnhetFeil
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.IngenNæringException
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.publiseringsdatoer.api.PubliseringsdatoerService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.aggregert.AggregertStatistikkDto
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.aggregert.AggregertStatistikkService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.kvartalsvis.KvartalsvisSykefraværshistorikk
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.kvartalsvis.KvartalsvisSykefraværshistorikkService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertLegemeldtSykefraværService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertSykefraværService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SummertSykefraværshistorikk
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.tilgangskontroll.TilgangskontrollException
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.tilgangskontroll.TilgangskontrollService
import no.nav.security.token.support.core.api.Protected
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*

@Protected
@RestController
class SykefraværshistorikkController(
    private val kvartalsvisSykefraværshistorikkService: KvartalsvisSykefraværshistorikkService,
    private val tilgangskontrollService: TilgangskontrollService,
    private val enhetsregisteretClient: EnhetsregisteretClient,
    private val summertSykefraværService: SummertSykefraværService,
    private val aggregertHistorikkService: AggregertStatistikkService,
    private val publiseringsdatoerService: PubliseringsdatoerService,
    private val summertLegemeldtSykefraværService: SummertLegemeldtSykefraværService
) {
    private val log: Logger = LoggerFactory.getLogger(this::class.java)

    @GetMapping(value = ["/{orgnr}/sykefravarshistorikk/kvartalsvis"])
    fun hentSykefraværshistorikk(
        @PathVariable("orgnr") orgnrStr: String, request: HttpServletRequest
    ): ResponseEntity<List<KvartalsvisSykefraværshistorikk>> {
        val orgnr = Orgnr(orgnrStr)
        val bruker = tilgangskontrollService.hentBrukerKunIaRettigheter()
        tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
            orgnr, bruker, request.method, "" + request.requestURL
        )
        val underenhet = enhetsregisteretClient.hentUnderenhet(orgnr)
        val overordnetEnhet = enhetsregisteretClient
            .hentEnhet(underenhet.overordnetEnhetOrgnr!!)
            .getOrElse { _: HentEnhetFeil ->
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
            }

        val harTilgangTilOverordnetEnhet =
            tilgangskontrollService.hentTilgangTilOverordnetEnhetOgLoggSikkerhetshendelse(
                bruker, overordnetEnhet, underenhet, request.method, "" + request.requestURL
            )
        return if (harTilgangTilOverordnetEnhet) {
            ResponseEntity.ok(
                kvartalsvisSykefraværshistorikkService.hentSykefraværshistorikk(
                    underenhet, overordnetEnhet
                )
            )
        } else {
            ResponseEntity.ok(
                kvartalsvisSykefraværshistorikkService.hentSykefraværshistorikk(
                    underenhet, overordnetEnhet.institusjonellSektorkode
                )
            )
        }
    }

    // TODO: Fjern har vi har gått over til "aggregert"-endepunktet
    @GetMapping(value = ["/{orgnr}/sykefravarshistorikk/summert"])
    fun hentSummertKorttidsOgLangtidsfraværV2(
        @PathVariable("orgnr") orgnrStr: String,
        @RequestParam("antallKvartaler") antallKvartaler: Int,
        request: HttpServletRequest
    ): List<SummertSykefraværshistorikk> {
        val bruker = tilgangskontrollService.hentBrukerKunIaRettigheter()
        tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
            Orgnr(orgnrStr), bruker, request.method, "" + request.requestURL
        )
        val underenhet = enhetsregisteretClient.hentUnderenhet(Orgnr(orgnrStr))
        require(antallKvartaler == 4) { "For øyeblikket støtter vi kun summering av 4 kvartaler." }
        val summertSykefraværshistorikkVirksomhet = summertSykefraværService.hentSummertSykefraværshistorikk(
            underenhet, publiseringsdatoerService.hentSistePubliserteKvartal(), antallKvartaler
        )
        val summertSykefraværshistorikkBransjeEllerNæring =
            summertSykefraværService.hentSummertSykefraværshistorikkForBransjeEllerNæring(
                underenhet, antallKvartaler
            )
        return listOf(
            summertSykefraværshistorikkVirksomhet, summertSykefraværshistorikkBransjeEllerNæring
        )
    }

    // TODO: Fjern etter at MFA har gått over til kafka
    @GetMapping(value = ["/{orgnr}/sykefravarshistorikk/legemeldtsykefravarsprosent"])
    fun hentLegemeldtSykefraværsprosent(
        @PathVariable("orgnr") orgnrStr: String, request: HttpServletRequest
    ): ResponseEntity<LegemeldtSykefraværsprosent?> {
        val orgnr = Orgnr(orgnrStr)
        val brukerMedIaRettigheter = tilgangskontrollService.hentBrukerKunIaRettigheter()
        val brukerMedMinstEnRettighet = tilgangskontrollService.hentInnloggetBrukerForAlleRettigheter()
        val brukerHarSykefraværRettighetTilVirksomhet = brukerMedIaRettigheter.harTilgang(orgnr)
        val brukerRepresentererVirksomheten = tilgangskontrollService.brukerRepresentererVirksomheten(orgnr)
        tilgangskontrollService.sjekkTilgangTilOrgnrOgLoggSikkerhetshendelse(
            Orgnr(orgnrStr),
            brukerMedMinstEnRettighet,
            request.method,
            "" + request.requestURL
        )
        val underenhet: Underenhet = try {
            enhetsregisteretClient.hentUnderenhet(orgnr)
        } catch (e: IngenNæringException) {
            return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null)
        }
        if (brukerHarSykefraværRettighetTilVirksomhet) {
            val legemeldtSykefraværsprosent = summertLegemeldtSykefraværService.hentLegemeldtSykefraværsprosent(
                underenhet, publiseringsdatoerService.hentSistePubliserteKvartal()
            )
            return if (legemeldtSykefraværsprosent.prosent == null) {
                ResponseEntity.status(HttpStatus.NO_CONTENT).body(null)
            } else ResponseEntity.status(HttpStatus.OK).body(legemeldtSykefraværsprosent)
        } else if (brukerRepresentererVirksomheten) {
            val legemeldtSykefraværsprosentBransjeEllerNæring = summertLegemeldtSykefraværService
                .hentLegemeldtSykefraværsprosentUtenStatistikkForVirksomhet(
                    underenhet, publiseringsdatoerService.hentSistePubliserteKvartal()
                )
            return if (legemeldtSykefraværsprosentBransjeEllerNæring.prosent == null) {
                ResponseEntity.status(HttpStatus.NO_CONTENT).body(null)
            } else ResponseEntity.status(HttpStatus.OK)
                .body(legemeldtSykefraværsprosentBransjeEllerNæring)
        }
        log.error("Brukeren har ikke tilgang til virksomhet, men ble ikke stoppet av tilgangssjekk.")
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(null)
    }

    @GetMapping("/{orgnr}/v1/sykefravarshistorikk/aggregert")
    fun hentAggregertStatistikk(
        @PathVariable("orgnr") orgnr: String
    ): ResponseEntity<AggregertStatistikkDto> {
        val statistikk = aggregertHistorikkService.hentAggregertStatistikk(Orgnr(orgnr))
            .getOrElseThrow { e: TilgangskontrollException? -> e }
        return ResponseEntity.status(HttpStatus.OK).body(statistikk)
    }
}